pkg/session/txn.go

```diff
@@ -683,29 +683,36 @@ func (txnFailFuture) Wait() (uint64, error) {
 
 // txnFuture is a promise, which promises to return a txn in future.
 type txnFuture struct {
-	future   oracle.Future
-	store    kv.Storage
-	txnScope string
+	future    oracle.Future
+	store     kv.Storage
+	txnScope  string
+	pipelined bool
 }
 
 func (tf *txnFuture) wait() (kv.Transaction, error) {
 	startTS, err := tf.future.Wait()
 	failpoint.Inject("txnFutureWait", func() {})
 	if err == nil {
+		if tf.pipelined {
+			return tf.store.Begin(tikv.WithTxnScope(tf.txnScope), tikv.WithStartTS(startTS), tikv.WithPipelinedMemDB())
+		}
 		return tf.store.Begin(tikv.WithTxnScope(tf.txnScope), tikv.WithStartTS(startTS))
 	} else if config.GetGlobalConfig().Store == "unistore" {
 		return nil, err
 	}
 
 	logutil.BgLogger().Warn("wait tso failed", zap.Error(err))
 	// It would retry get timestamp.
+	if tf.pipelined {
+		return tf.store.Begin(tikv.WithTxnScope(tf.txnScope), tikv.WithPipelinedMemDB())
+	}
 	return tf.store.Begin(tikv.WithTxnScope(tf.txnScope))
 }
 
 // HasDirtyContent checks whether there's dirty update on the given table.
 // Put this function here is to avoid cycle import.
 func (s *session) HasDirtyContent(tid int64) bool {
-	if s.txn.Transaction == nil {
+	if s.txn.Transaction == nil || s.txn.Transaction.IsPipelined() {
```

Comment:
Why put this? If it has to, can we add comments explaining it?

Reply:
Comment added.